app/viewModels/houseConsignment/index/items/packages/AddAnotherPackageViewModel.scala

```diff
@@ -0,0 +1,111 @@
+ */
+
+package viewModels.houseConsignment.index.items.packages
+
+import config.FrontendAppConfig
+import models.{ArrivalId, Index, Mode, RichOptionalJsArray, UserAnswers}
+import pages.houseConsignment.index.items.packages.{NumberOfPackagesPage, PackageTypePage}
+import pages.sections.PackagingListSection
+import play.api.i18n.Messages
+import play.api.libs.json.JsArray
+import play.api.mvc.Call
+import viewModels.{AddAnotherViewModel, ListItem}
+
+case class AddAnotherPackageViewModel(
+  listItems: Seq[ListItem],
+  onSubmitCall: Call,
+  nextIndex: Index
+) extends AddAnotherViewModel {
+  override val prefix: String = "houseConsignment.index.items.packages.addAnotherPackage"
+
+  override def maxCount(implicit config: FrontendAppConfig): Int = config.maxPackages
+
+  def title(houseConsignmentIndex: Index, itemIndex: Index)(implicit messages: Messages): String =
+    messages(s"$prefix.$emptyOrSingularOrPlural.title", count, itemIndex.display, houseConsignmentIndex.display)
+
+  def heading(houseConsignmentIndex: Index, itemIndex: Index)(implicit messages: Messages): String =
+    messages(s"$prefix.$emptyOrSingularOrPlural.heading", count, itemIndex.display, houseConsignmentIndex.display)
+
+  def legend(houseConsignmentIndex: Index, itemIndex: Index)(implicit messages: Messages): String =
+    if (count <= 0) {
+      messages(s"$prefix.empty.label", count, itemIndex.display, houseConsignmentIndex.display)
+    } else {
+      messages(s"$prefix.label", count, itemIndex.display, houseConsignmentIndex.display)
+    }
+
+  def maxLimitLabel(houseConsignment: Index, itemIndex: Index)(implicit messages: Messages): String =
+    messages(s"$prefix.maxLimit.label", count, itemIndex.display, houseConsignment.display)
+
+}
+
+object AddAnotherPackageViewModel {
+
+  class AddAnotherPackageViewModelProvider {
+
+    def apply(userAnswers: UserAnswers, arrivalId: ArrivalId, houseConsignmentIndex: Index, itemIndex: Index, mode: Mode): AddAnotherPackageViewModel = {
+
+      val array = userAnswers.get(PackagingListSection(houseConsignmentIndex, itemIndex))
+
+      val listItems = array
+        .getOrElse(JsArray())
+        .value
+        .zipWithIndex
+        .flatMap {
+          case (_, index) =>
+            val packageIndex = Index(index)
+
+            userAnswers
+              .get(PackageTypePage(houseConsignmentIndex, itemIndex, packageIndex))
+              .flatMap(
+                packageType =>
+                  userAnswers
+                    .get(NumberOfPackagesPage(houseConsignmentIndex, itemIndex, packageIndex))
+                    .map(
+                      quantity => s"$quantity * $packageType"
+                    )
+              )
+              .orElse(
+                userAnswers
+                  .get(PackageTypePage(houseConsignmentIndex, itemIndex, packageIndex))
+                  .map(
+                    packageType => s"$packageType"
+                  )
+              )
```

Comment:
I imagine this can be refactored to something like

```
      val listItems = array
        .flatMapWithIndex {
          case (_, packageIndex) =>
            userAnswers
              .get(PackageTypePage(houseConsignmentIndex, itemIndex, packageIndex))
              .flatMap(
                packageType =>
                  userAnswers
                    .get(NumberOfPackagesPage(houseConsignmentIndex, itemIndex, packageIndex))
                    .map {
                      quantity => s"$quantity * $packageType"
                    }
                    .orElse {
                      Some(s"$packageType")
                    }
              )
              .map {
                name =>
                  ListItem(
                    name = name,
                    changeUrl = None,
                    removeUrl = Some(
                      controllers.houseConsignment.index.items.packages.routes.RemovePackageTypeYesNoController
                        .onPageLoad(arrivalId, houseConsignmentIndex, itemIndex, packageIndex, mode)
                        .url
                    )
                  )
              }
        }
```

Reply:
Which can be refactored further to

```
      val listItems = array
        .flatMapWithIndex {
          case (_, packageIndex) =>
            userAnswers
              .get(PackageTypePage(houseConsignmentIndex, itemIndex, packageIndex))
              .map {
                packageType =>
                  userAnswers
                    .get(NumberOfPackagesPage(houseConsignmentIndex, itemIndex, packageIndex))
                    .fold {
                      s"$packageType"
                    } {
                      quantity => s"$quantity * $packageType"
                    }
              }
              .map {
                name =>
                  ListItem(
                    name = name,
                    changeUrl = None,
                    removeUrl = Some(
                      controllers.houseConsignment.index.items.packages.routes.RemovePackageTypeYesNoController
                        .onPageLoad(arrivalId, houseConsignmentIndex, itemIndex, packageIndex, mode)
                        .url
                    )
                  )
              }
        }
```